pkg/executor/join/hash_join_base.go

```diff
@@ -218,15 +238,44 @@ type buildWorkerBase struct {
 	BuildKeyColIdx []int
 }
 
-// fetchBuildSideRows fetches all rows from build side executor, and append them
-// to e.buildSideResult.
+func syncerAdd(syncer *sync.WaitGroup) {
+	if syncer != nil {
+		syncer.Add(1)
+	}
+}
+
+func syncerDone(syncer *sync.WaitGroup) {
+	if syncer != nil {
+		syncer.Done()
+	}
+}
+
 func (w *buildWorkerBase) fetchBuildSideRows(ctx context.Context, hashJoinCtx *hashJoinCtxBase, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
-	defer close(chkCh)
-	var err error
+	w.fetchBuildSideRowsImpl(ctx, hashJoinCtx, nil, chkCh, errCh, doneCh)
+}
+
+// fetchBuildSideRowsImpl fetches all rows from build side executor, and append them
+// to e.buildSideResult.
+func (w *buildWorkerBase) fetchBuildSideRowsImpl(ctx context.Context, hashJoinCtx *hashJoinCtxBase, fetcherAndWorkerSyncer *sync.WaitGroup, chkCh chan<- *chunk.Chunk, errCh chan<- error, doneCh <-chan struct{}) {
```

Comment:
why not add `fetcherAndWorkerSyncer` as an argument of `fetchBuildSideRows` and let the caller to pass this argument

Reply:
> why not add `fetcherAndWorkerSyncer` as an argument of `fetchBuildSideRows` and let the caller to pass this argument

okk